### PR TITLE
feat: Add prometheus deployment manifests to setup tools

### PR DESCRIPTION
**Pull request by bradbeam**
Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

### CODE REVIEW

File: config/docker-entrypoint.sh [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
+  ;;
+  chart)
+    # Package the Helm chart
     shift && /workspace/chart/build.sh "$@"
     exit $?
-fi
+  ;;
+  prometheus)
+    # Generate prometheus manifests
+    shift && cd /workspace/rso
```
Comment (jgustie) [inline config/docker-entrypoint.sh]: This should be `/workspace/prometheus` you can just wipe out the existing service monitor (the endpoints it would have been using are gone, it's dead code).

File: config/docker-entrypoint.sh [hunk 2]
```diff
@@ -31,6 +34,8 @@ fi
 
 # Add trial labels to the resulting manifests so they can be more easily located
 if [ -n "$TRIAL" ]; then
+    # Note, this heredoc block must be indented with tabs
+    # <<- allows for indentation via tabs, if spaces are used it is no good.
```
Comment (jgustie) [inline config/docker-entrypoint.sh:46]: I'm sensing you got burned by this?

File: config/docker-entrypoint.sh [hunk 3]
```diff
@@ -44,12 +49,29 @@ if [ -n "$TRIAL" ]; then
...
+    cat <<-EOF >"experiment_labels.yaml"
+		apiVersion: konjure.carbonrelay.com/v1beta1
+		kind: LabelTransformer
+		metadata:
+		  name: experiment-labels
+		labels:
+		  "redskyops.dev/experiment": $EXPERIMENT
+		  "redskyops.dev/experiment": experimentResource
```
Comment (jgustie) [inline config/docker-entrypoint.sh]: This second (duplicate) label should be `experiment-role`; also you should add it to `well_known_labels.go`.

File: config/docker-entrypoint.sh [hunk 4]
```diff
@@ -44,12 +49,29 @@ if [ -n "$TRIAL" ]; then
...
+fi
+
+
 # Process arguments
 while [ "$#" != "0" ] ; do
     case "$1" in
     create)
         handle () {
-            kubectl create -f -
+            kubectl apply -f -
```
Comment (jgustie) [inline config/docker-entrypoint.sh]: This was explicitly `create` for `generate-name` and to force failure on conflict. You should add another case for `apply` (it used to be there but got removed after we weren't using it for anything).

File: config/rso/kube-state-metrics-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,52 @@
+---
+# Source: prometheus/charts/kube-state-metrics/templates/deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: kube-state-metrics
+  namespace: default
```
Comment (jgustie) [inline config/rso/kube-state-metrics-deployment.yaml]: We shouldn't be putting things in the default namespace like this: I would have expected the namespace to be omitted from all of the resources (even if the edit at runtime is going to add a namespace transformer).

Also, even though the entry point script should create the right kustomization by scanning for resources, it should still be explicitly included in the directory and should still explicitly list all of the resources).

File: config/rso/prometheus-server-rbac.yaml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+#      - ingresses
+#    verbs:
+#      - get
+#      - list
+#      - watch
+#  - nonResourceURLs:
+#      - "/metrics"
+#    verbs:
+#      - get
+#
```
Comment (jgustie) [inline config/rso/prometheus-server-rbac.yaml]: Remove these comments, do it in a separate commit if you might want them back.

**Review by jgustie — CHANGES_REQUESTED** [review]

File: config/docker-entrypoint.sh (hunk 2 above)
Comment (bradbeam) [reply config/docker-entrypoint.sh:46]: yup. saw the indentation looked off, #thereifixedit and it no longer worked.

File: config/prometheus/prometheus-server-service.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
+---
+# Source: prometheus/templates/services/server.yaml
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app: prometheus
+  name: prometheus-server
```
Comment (jgustie) [inline config/prometheus/prometheus-server-service.yaml]: Can we just name this `prometheus` so when we have URL support we can have a default of `http://prometheus:9090` ?

File: config/prometheus/prometheus-server-service.yaml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+spec:
+  ports:
+  - name: http
+    port: 9090
+    protocol: TCP
+    targetPort: 9090
+  selector:
+    app: prometheus
+  sessionAffinity: None
+  type: ClusterIP
```
Comment (jgustie) [inline config/prometheus/prometheus-server-service.yaml]: Is this a duplicate service?

File: config/prometheus/prometheus-server-deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+          args:
+            - --volume-dir=/etc/config
+            - --webhook-url=http://127.0.0.1:9090/-/reload
+          volumeMounts:
+            - name: config-volume
+              mountPath: /etc/config
+              readOnly: true
+        - name: prometheus-server
+          image: "prom/prometheus:latest"
+          imagePullPolicy: "IfNotPresent"
```
Comment (jgustie) [inline config/prometheus/prometheus-server-deployment.yaml]: I would be a little concerned trusting `latest` here, can we pin this to a version (at least a major, maybe major/minor)?

**Review by jgustie — DISMISSED** [review]

File: config/prometheus/prometheus-server-service.yaml (hunk 7 above)
Comment (bradbeam) [reply config/prometheus/prometheus-server-service.yaml]: Want to avoid `prometheus` since that has a potential for conflict. Will use `rso-prometheus` below and remove this one.

Comment (bradbeam) [reply config/prometheus/prometheus-server-service.yaml]: along that vein I should probably change all the basic 'prometheus' names to something more specific. ugh.

Comment (jgustie) [reply config/prometheus/prometheus-server-service.yaml]: I like the change of the name prefix for all the other objects, I still don't like it on the service because of the whole sensible defaults thing.

A default of "http://prometheus:9090" is a good default because it is likely to actually hit (our Prometheus or theirs, it's generic); if the service name is "redsky-prometheus" then we either need to make the default URL "http://redsky-prometheus:9090" (only likely to hit our Prometheus) or stick with "http://prometheus:9090" and have a default that needs to be overwritten in the particular use case we established the default behavior for in the first place.

I'm fine to revisit this once we are closer with URL based metric collection.

**Review by jgustie — APPROVED** [review]
